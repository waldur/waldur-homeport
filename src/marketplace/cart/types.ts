import { OrderItemResponse } from '@waldur/marketplace/orders/types';
import { AttributesType, Offering, Plan } from '@waldur/marketplace/types';

export interface State {
  items: OrderItemResponse[];
  addingItem: boolean;
  removingItem: boolean;
  updatingItem: boolean;
  creatingOrder: boolean;
}

export interface OuterState {
  marketplace: {
    cart: State;
  };
}

export interface SubmitCartRequest {
  project: string;
}

export interface OrderItemRequest {
  offering: Offering;
  plan?: Plan;
  attributes?: AttributesType;
  limits?: { [key: string]: number };
}
