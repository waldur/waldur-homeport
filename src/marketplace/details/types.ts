import { Offering, Plan } from '@waldur/marketplace/types';
import { Customer, Project } from '@waldur/workspace/types';

export interface Limits {
  [key: string]: number;
}

export interface OfferingFormData {
  plan?: Plan;
  attributes?: { [key: string]: any };
  project?: Project;
  limits?: Limits;
}

export interface OrderSummaryProps {
  offering: Offering;
  customer: Customer;
  project?: Project;
  total: number;
  formData: OfferingFormData;
  formValid: boolean;
  updateMode?: boolean;
  extraComponent?: React.ComponentType<any>;
}
