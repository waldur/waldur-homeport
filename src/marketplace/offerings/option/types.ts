export type FieldType =
  | 'boolean'
  | 'integer'
  | 'money'
  | 'string'
  | 'text'
  | 'html_text'
  | 'select_string'
  | 'select_openstack_tenant'
  | 'select_openstack_instance'
  | 'select_multiple_openstack_instances'
  | 'date'
  | 'time';
