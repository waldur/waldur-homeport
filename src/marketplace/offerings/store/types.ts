import { Option } from '@waldur/marketplace/common/registry';
import {
  AttributesType,
  Category,
  OfferingComponent,
  OfferingOptions,
} from '@waldur/marketplace/types';

export interface PlanFormData {
  archived: boolean;
  name: string;
  unit: Option;
  unit_price: number;
  prices: { [key: string]: number };
  quotas: { [key: string]: number };
  description?: string;
  article_code?: string;
  uuid?: string;
}

export interface OptionFormData {
  name: string;
  label: string;
  type: Option;
  choices: string;
}

export interface ScheduleFormData {
  start_date: string;
  start_time: string;
  end_date: string;
  end_time: string;
  type: string;
  title: string;
  id: string;
}

export type OfferingLimits = Record<string, { min: number; max: number }>;

export interface OfferingFormData {
  name: string;
  description?: string;
  full_description?: string;
  terms_of_service?: string;
  category: Category;
  type: Option;
  attributes?: AttributesType;
  components?: OfferingComponent[];
  plans?: PlanFormData[];
  options?: OptionFormData[];
  schedules?: ScheduleFormData[];
  plugin_options?: Record<string, any>;
  secret_options?: Record<string, any>;
  service_settings?: any;
  thumbnail?: File;
  scope?: string;
  document?: OfferingDocument;
  limits?: OfferingLimits;
}

export interface OfferingUpdateFormData extends OfferingFormData {
  offeringUuid: string;
}

export interface PlanRequest {
  name: string;
  unit: string;
  unit_price: number;
  quotas?: { [key: string]: number };
  prices?: { [key: string]: number };
  description?: string;
  article_code?: string;
  uuid?: string;
}

export interface OfferingRequest {
  name: string;
  description?: string;
  full_description?: string;
  terms_of_service?: string;
  type: string;
  customer: string;
  category: string;
  attributes?: AttributesType;
  components?: OfferingComponent[];
  plans?: PlanRequest[];
  options?: OfferingOptions;
  plugin_options?: Record<string, any>;
  secret_options?: Record<string, any>;
  schedules?: ScheduleFormData[];
  scope?: string;
  service_attributes?: any;
  shared: boolean;
  limits?: OfferingLimits;
}

export interface OfferingDocument {
  file: File;
  name?: string;
}
