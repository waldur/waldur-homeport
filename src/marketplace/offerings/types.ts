export type OfferingStep =
  | 'Overview'
  | 'Description'
  | 'Management'
  | 'Accounting'
  | 'Review';

export const STEPS: OfferingStep[] = [
  'Overview',
  'Description',
  'Management',
  'Accounting',
  'Review',
];

export interface FilterQuery {
  name?: string;
  attributes?: object;
}

export interface Choice {
  value: string;
  label: string;
}
