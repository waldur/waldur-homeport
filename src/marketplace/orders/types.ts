import { AttributesType, BillingPeriod } from '@waldur/marketplace/types';

export type OrderStep = 'Configure' | 'Approve' | 'Review';

export type OrderState =
  | 'requested for approval'
  | 'executing'
  | 'done'
  | 'terminated'
  | 'erred';

export interface StatusChange {
  approving: boolean;
  rejecting: boolean;
}

export type OrderItemType = 'Create' | 'Update' | 'Terminate';

export interface OrderItemResponse {
  name?: string;
  uuid: string;
  type: OrderItemType;
  offering: string;
  offering_uuid: string;
  offering_name: string;
  offering_description: string;
  offering_thumbnail: string;
  offering_type: string;
  offering_terms_of_service: string;
  offering_shared: boolean;
  offering_billable: boolean;
  marketplace_resource_uuid?: string;
  resource_name?: string;
  resource_uuid?: string;
  resource_type?: string;
  cost: string;
  estimate?: number;
  unit: string;
  state: string;
  attributes: AttributesType;
  plan_uuid?: string;
  plan?: string;
  plan_name?: string;
  plan_description?: string;
  plan_unit?: BillingPeriod;
  project: string;
  project_name: string;
  customer_name?: string;
  customer_uuid?: string;
  category_title?: string;
  category_uuid?: string;
  created: string;
  error_message?: string;
  limits: Record<string, number>;
  current_usages?: Record<string, number>;
  scope?: string;
}

export interface OrderItemDetailsType extends OrderItemResponse {
  order_uuid: string;
  order_approved_at?: string;
  order_approved_by?: string;
  created_by_full_name: string;
  created_by_civil_number: string;
  customer_name: string;
  customer_uuid: string;
  project_name: string;
  project_uuid: string;
  old_plan_name: string;
  new_plan_name: string;
  old_cost_estimate: string;
  new_cost_estimate: string;
  can_terminate: boolean;
}

export interface Order {
  items: OrderItemResponse[];
  state: OrderState;
  total_cost?: number;
  file?: string;
  project?: string;
  project_uuid?: string;
}
