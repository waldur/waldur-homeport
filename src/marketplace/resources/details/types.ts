import { StateIndicatorProps } from '@waldur/core/StateIndicator';

interface DataRow {
  name: string;
  state?: StateIndicatorProps;
  summary: string;
  marketplace_resource_uuid?: string;
  url?: string;
  project_uuid?: string;
}

export interface DataPage {
  nextPage: number;
  data: DataRow[];
}
