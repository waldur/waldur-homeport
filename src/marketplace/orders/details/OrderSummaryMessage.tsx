import { DateTime } from 'luxon';
import { FunctionComponent, useMemo } from 'react';

import { defaultCurrency } from '@waldur/core/formatCurrency';
import { translate } from '@waldur/i18n';
import { OrderDetailsType } from '@waldur/marketplace/orders/types';
import { Offering, OfferingComponent } from '@waldur/marketplace/types';

interface OrderSummaryProps {
  order: OrderDetailsType;
  offering: Offering;
}

type ComponentMap = Record<string, OfferingComponent>;

type Limits = Record<string, number>;

interface Context {
  order: Pick<
    OrderDetailsType,
    | 'attributes'
    | 'limits'
    | 'resource_name'
    | 'old_plan_name'
    | 'new_plan_name'
    | 'old_cost_estimate'
    | 'new_cost_estimate'
  >;
  user: string;
  approved?: string;
  components: OfferingComponent[];
}

const getCreateSummary = (ctx: Context): string => {
  const msg = translate(
    '{user} has requested provisioning of "{resource_name}" with plan "{plan_name}".',
    {
      user: ctx.user,
      resource_name: ctx.order.attributes.name,
      plan_name: ctx.order.new_plan_name || 'Default',
    },
  );
  if (ctx.approved) {
    return msg + ' ' + ctx.approved;
  } else {
    return msg;
  }
};

const formatComponent = (
  limits: Limits,
  componentMap: ComponentMap,
  key: string,
) =>
  [
    `${key.toLocaleUpperCase()}:`,
    limits[key] / (componentMap[key]?.factor || 1),
    componentMap[key]?.measured_unit,
  ]
    .filter((x) => x)
    .join(' ');

const formatLimits = (limits: Limits, componentMap: ComponentMap) =>
  Object.keys(limits)
    .sort()
    .map((key) => formatComponent(limits, componentMap, key))
    .join(', ');

const getComponentMap = (components: OfferingComponent[]): ComponentMap =>
  components.reduce((r, c) => ({ ...r, [c.type]: c }), {});

export const getUpdateSummary = (ctx: Context) => {
  let msg;
  const componentMap = getComponentMap(ctx.components);
  if (ctx.order.attributes.old_limits) {
    msg = translate(
      '{user} has requested changing of limits from "{old_limits}" to "{new_limits}".',
      {
        user: ctx.user,
        old_limits: formatLimits(ctx.order.attributes.old_limits, componentMap),
        new_limits: formatLimits(ctx.order.limits, componentMap),
      },
    );
  } else {
    msg = translate(
      '{user} has requested changing of plan from "{old_plan}" to "{new_plan}".',
      {
        user: ctx.user,
        old_plan: ctx.order.old_plan_name || 'Default',
        new_plan: ctx.order.new_plan_name,
      },
    );
  }
  if (ctx.approved) {
    return msg + ' ' + ctx.approved;
  } else {
    return (
      msg +
      ' ' +
      translate(
        'Estimated monthly fee will change from {old_estimate} to {new_estimate}. VAT is not included.',
        {
          old_estimate: defaultCurrency(ctx.order.old_cost_estimate || 0),
          new_estimate: defaultCurrency(ctx.order.new_cost_estimate || 0),
        },
      )
    );
  }
};

const getTerminateSummary = (ctx: Context) => {
  const msg = translate(
    '{user} has requested termination of {resource_name} with active plan "{plan_name}".',
    {
      user: ctx.user,
      resource_name: ctx.order.resource_name,
      plan_name: ctx.order.old_plan_name || 'Default',
    },
  );
  if (ctx.approved) {
    return msg + ' ' + ctx.approved;
  } else {
    return msg;
  }
};

const getContext = (order: OrderDetailsType, offering: Offering): Context => {
  let user = order.created_by_full_name;
  if (order.created_by_civil_number) {
    user += ` (ID: ${order.created_by_civil_number})`;
  }
  let approved;
  if (order.consumer_reviewed_at && order.consumer_reviewed_by) {
    approved = translate('{user} has approved it on {date} at {time}.', {
      user: order.consumer_reviewed_by,
      date: DateTime.fromISO(order.consumer_reviewed_at).toISODate(),
      time: DateTime.fromISO(order.consumer_reviewed_at).toFormat('HH:mm'),
    });
  }
  return { user, approved, order, components: offering.components };
};

const getMessage = (order: OrderDetailsType, offering: Offering): string => {
  const ctx = getContext(order, offering);
  if (order.type === 'Create') {
    return getCreateSummary(ctx);
  } else if (order.type === 'Update') {
    return getUpdateSummary(ctx);
  } else if (order.type === 'Terminate') {
    return getTerminateSummary(ctx);
  }
};

export const OrderSummaryMessage: FunctionComponent<OrderSummaryProps> = (
  props,
) => {
  const message = useMemo(
    () => getMessage(props.order, props.offering),
    [props.order, props.offering],
  );
  return <>{message}</>;
};
