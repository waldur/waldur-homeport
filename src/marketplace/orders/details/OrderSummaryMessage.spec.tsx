import { BillingType } from '@waldur/marketplace/types';

import { getUpdateSummary } from './OrderSummaryMessage';

jest.mock('@waldur/core/formatCurrency', () => ({
  defaultCurrency: (val) => val,
}));

const COMPONENTS = [
  {
    billing_type: 'usage' as BillingType,
    type: 'gigabytes_gpfs',
    name: 'Storage (gpfs)',
    description: 'Capacity optimized HDD',
    measured_unit: 'GB',
    limit_period: null,
    limit_amount: null,
    article_code: '',
    max_value: null,
    min_value: null,
    factor: null,
    uuid: '10ff42e159d7472fb35d4db517e6c16e',
  },
  {
    billing_type: 'usage' as BillingType,
    type: 'gigabytes_rbd-ec',
    name: 'Storage (rbd-ec)',
    description: 'General purpose SSD',
    measured_unit: 'GB',
    limit_period: null,
    limit_amount: null,
    article_code: '',
    max_value: null,
    min_value: null,
    factor: null,
    uuid: '36d5008a3b6611eebe560242ac120002',
  },
  {
    billing_type: 'fixed' as BillingType,
    type: 'cores',
    name: 'Cores',
    description: '',
    measured_unit: 'cores',
    limit_period: null,
    limit_amount: null,
    article_code: '',
    max_value: 500,
    min_value: 1,
    factor: 1,
    uuid: '36d506483b6611eebe560242ac120002',
  },
  {
    billing_type: 'fixed' as BillingType,
    type: 'ram',
    name: 'RAM',
    description: '',
    measured_unit: 'GB',
    limit_period: null,
    limit_amount: null,
    article_code: '',
    max_value: 20480000,
    min_value: 1024,
    factor: 1024,
    uuid: '36d508503b6611eebe560242ac120002',
  },
  {
    billing_type: 'fixed' as BillingType,
    type: 'storage',
    name: 'Storage',
    description: '',
    measured_unit: 'GB',
    limit_period: null,
    limit_amount: null,
    article_code: '',
    max_value: 102400000,
    min_value: 10240,
    factor: 1024,
    uuid: '36d50a8a3b6611eebe560242ac120002',
  },
];

describe('OrderSummary', () => {
  const order = {
    attributes: {
      old_limits: {
        cpu: 10,
        storage: 1024 * 100,
      },
    },
    limits: {
      cpu: 30,
      storage: 1024 * 300,
    },
    resource_name: 'Demo Cloud',
    old_plan_name: 'Basic',
    new_plan_name: 'Advanced',
    old_cost_estimate: '100',
    new_cost_estimate: '300',
  };

  it('formats update summary with components', () => {
    const ctx = {
      order,
      user: 'Alice Lebowski',
      components: COMPONENTS,
    };
    expect(getUpdateSummary(ctx)).toMatchSnapshot();
  });

  it('formats update summary without components', () => {
    const ctx = {
      order,
      user: 'Alice Lebowski',
      components: [],
    };
    expect(getUpdateSummary(ctx)).toMatchSnapshot();
  });
});
