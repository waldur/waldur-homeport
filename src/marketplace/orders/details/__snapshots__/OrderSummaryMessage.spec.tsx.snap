// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderSummary formats update summary with components 1`] = `"Alice Lebowski has requested changing of limits from \\"CPU: 10, STORAGE: 100 GB\\" to \\"CPU: 30, STORAGE: 300 GB\\". Estimated monthly fee will change from 100 to 300. VAT is not included."`;

exports[`OrderSummary formats update summary without components 1`] = `"Alice Lebowski has requested changing of limits from \\"CPU: 10, STORAGE: 102400\\" to \\"CPU: 30, STORAGE: 307200\\". Estimated monthly fee will change from 100 to 300. VAT is not included."`;
